python/neuroglancer/pipeline/volumes/cloudvolume.py

```diff
@@ -0,0 +1,486 @@
+from __future__ import print_function
+
+from collections import namedtuple
+import json
+import os
+import re
+
+import numpy as np
+from tqdm import tqdm
+
+import neuroglancer
+from neuroglancer.lib import clamp, xyzrange, Vec, Bbox, min2, max2, find_closest_divisor
+from volumes import Volume, VolumeCutout, generate_slices
+from neuroglancer.pipeline.storage import Storage
+
+__all__ = [ 'CloudVolume', 'EmptyVolumeException' ]
+
+ExtractedPath = namedtuple('ExtractedPath', 
+  ('protocol','bucket_name', 'dataset_name','layer_name')
+)
+
+class EmptyVolumeException(Exception):
+    pass
+
+class CloudVolume(Volume):
+  """
+  CloudVolume represents an interface to a dataset layer at a given
+  mip level. You can use it to send and receive data from neuroglancer
+  datasets on supported hosts like Google Storage, S3, and local Filesystems.
+
+  Required:
+    cloudpath: Path to the dataset layer. This should match storage's supported
+      providers.
+
+      e.g. Google: gs://neuroglancer/$DATASET/$LAYER/
+           S3    : s3://neuroglancer/$DATASET/$LAYER/
+           Lcl FS: file:///tmp/$DATASET/$LAYER/
+  Optional:
+    mip: (int) Which level of downsampling to read to/write from. 0 is the highest resolution.
+    bounded: (bool) If a region outside of volume bounds is accessed:
+        True: 
+          - Throw an error
+          - Negative indicies have the normal python meaning
+        False: 
+          - Fill the region with black (useful for e.g. marching cubes's 1px boundary)
+          - Negative indicies refer to cartesian space
+    fill_missing: (bool) If a file inside volume bounds is unable to be fetched:
+        True: Use a block of zeros
+        False: Throw an error
+    info: (dict) in lieu of fetching a neuroglancer info file, use this provided one.
+            This is useful when creating new datasets.
+  """
+  def __init__(self, cloudpath, mip=0, bounded=True, fill_missing=False, info=None):
+    super(self.__class__, self).__init__()
+
+    extract = CloudVolume.extract_path(cloudpath)
+
+    self._protocol = extract.protocol
+    self._bucket = extract.bucket_name
+
+    # You can access these two with properties
+    self._dataset_name = extract.dataset_name
+    self._layer = extract.layer_name
+
+    self.mip = mip
+    self.bounded = bounded
+    self.fill_missing = fill_missing
+
+    self._storage = Storage(self.layer_cloudpath, n_threads=0)
+
+    if info is None:
+      self.refreshInfo()
+    else:
+      self.info = info
+
+    try:
+      self.mip = self.available_mips[self.mip]
+    except:
+      raise Exception("MIP {} has not been generated.".format(self.mip))
+
+  @classmethod
+  def create_new_info(cls, num_channels, layer_type, data_type, encoding, resolution, voxel_offset, volume_size, mesh=None, chunk_size=[64,64,64]):
```

Comment:
Yea it's fine. It's made so that you can specify the name of the mesh directory now instead of saying "it exists" or "it doesn't exist"